packages/dds/tree/src/changeset/delta.ts

```diff
@@ -0,0 +1,291 @@
+ * }
+ * export interface Delete {
+ *   [type]: typeof MarkType.Delete;
+ *   count: number;
+ * }
+ * export interface MoveOut {
+ *   [type]: typeof MarkType.MoveOut;
+ *   count: number;
+ *   moveId: MoveId;
+ * }
+ * ```
+ * Note the absence of modify information in `Delete` and `MoveOut` above.
+ *
+ * The benefit of the chosen representation over the alternative are two-fold:
+ * - It leads to less splitting of moved and deleted ranges of nodes.
+ * - It makes the format more uniform since modifications to moved-in subtrees must be represented with modifications
+ * marks within a `MoveIn` mark.
+ *
+ * 5. `MoveIn` marks are represented in the location where the content being moved resides in the input context that
+ * the delta is applied to, and `MoveOut` marks are represented in the location where the content being moved should
+ * reside after the delta is applied.
+ *
+ * The alternative would be to allow such marks to appear in temporary locations (e.g., in field "bar" for a
+ * transaction that moves content from "foo" to "bar" then moves that same content from "bar" to "baz").
+ * This makes the format less terse and harder to reason about.
+ *
+ * 6. MoveIn marks are not inlined within `ProtoFields`.
+ *
+ * Inlining them would force the consuming code to detect `MoveIn` marks within the `ProtoFields` and handle them
+ * within the context of the insert.
+ * This would be cumbersome because either the code that is responsible for consuming the `ProtoFields` would need to
+ * be aware of and have the context to handle `MoveIn`, or some caller of that code would need to find and extract such
+ * `MoveIn` marks ahead to calling that code.
+ */
+
+/**
+ * Represents the change made to a document.
+ */
+export type Root = PositionedMarks;
+
+/**
+ * Represents a list of changes to some range of nodes. The index of each mark within the range of nodes, before
+ * applying any of the changes, is not represented explicitly.
+ * It corresponds to the sum of the following:
+ * - The offsets of all preceding marks
+ * - The number of nodes affected by all preceding marks, this is computed as follows:
+ *   - Modify marks affect a single node
+ *   - Delete marks affect `Delete.count` nodes
+ *   - MoveOut marks affect `MoveOut.count` nodes
+ * - The offset of the given mark
+ */
+export type PositionedMarks<TMark = Mark> = MarkWithOffset<TMark>[];
+
+export const type: unique symbol = Symbol("Delta.type");
+export const setValue: unique symbol = Symbol("Delta.setValue");
+
+export type Mark = Modify | Delete | MoveOut | MoveIn | Insert;
+
+/**
+ * See PositionedMarks.
+ */
+export interface MarkWithOffset<TMark = Mark> {
+    offset: Offset;
+    mark: TMark;
+}
+
+/**
+ * Describes modifications made to a subtree that is otherwise untouched (i.e., not being inserted, deleted, or
+ * moved).
+ */
+export interface Modify {
+    [type]: typeof MarkType.Modify;
+    [setValue]?: Value;
+    [key: FieldKey]: PositionedMarks;
+}
```

Comment:
Whats the motivation for picking the symbols + field key indexing properts instead of normal properties and a separate Map<FieldKey, PositionedMarks>?

Since neither are json serializable, I'd normally go with the normal properties + map approach. I pretty much only use indexing of objects instead of maps for json compatibility.

Reply:
Note that using the map will let you actually use our banded FieldKey type here as well.

Reply:
The motivation was just to reduce the amount of nesting.